### PR TITLE
Value-set analysis: templatise and virtualise to facilitate customisation

### PR DESCRIPTION
**Pull request by smowton**
This provides sufficient hooks in value-set analysis, a CBMC core pointer-analysis, to permit downstream products such as the security analyser to create derivatives that share most of its logic.

VSA itself is currently unused, as revealed by the need to add support for DEAD statements before it can be used; this also adds unit tests to check its basic functionality has not broken in the meantime.

VSA is based on `static_analysist` and so will need to be ported to `ait` in the future if we wish to continue using it; however this should be orthogonal to permitting it to be customised like this.

### CODE REVIEW

File: src/pointer-analysis/value_set.h [hunk 1]
```diff
@@ -211,11 +211,11 @@ class value_sett
     const exprt &expr,
     const namespacet &ns);
 
-  void apply_code(
+  virtual void apply_code(
```
Comment (reuk) [inline src/pointer-analysis/value_set.h]: In general `public virtual` methods in non-abstract classes are a code smell, as they have two many responsibilities. They specify both the public interface, and a customization point for derived classes (which is an implementation detail). Ideally, the two responsibilities should be decoupled by providing a `private virtual` method which is called by a `public` (non-virtual) method. Then, if interface requirements change, the interface can be adjusted independently of the implementation (and vice versa).

I'm particularly concerned about retroactively making this method virtual. Pre-existing clients which always expect this method to do the same thing could be broken when derived classes change its behavior.

File: src/pointer-analysis/value_set.h [hunk 2]
```diff
@@ -242,28 +242,28 @@ class value_sett
     const namespacet &ns) const;
 
 protected:
```
Comment (reuk) [inline src/pointer-analysis/value_set.h:259]: Prefer to make these `private` if at all possible.

File: src/pointer-analysis/value_set_analysis.h [hunk 3]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+#include <util/xml.h>
 
 #include "value_set_domain.h"
 #include "value_sets.h"
+#include "value_set.h"
 
 class xmlt;
 
-class value_set_analysist:
+template<class Value_Sett>
```
Comment (reuk) [inline src/pointer-analysis/value_set_analysis.h]: Plz no camel snake case. `ValueSet`?

File: src/pointer-analysis/value_set_analysis.h [hunk 4]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  public static_analysist<value_set_domaint<Value_Sett> >
 {
 public:
-  explicit value_set_analysist(const namespacet &_ns):
-    static_analysist<value_set_domaint>(_ns)
+  typedef value_set_domaint<Value_Sett> domaint;
+  typedef static_analysist<domaint> baset;
+  typedef typename baset::locationt locationt;
+
+  explicit value_set_analysis_baset(const namespacet &_ns):baset(_ns)
```
Comment (reuk) [inline src/pointer-analysis/value_set_analysis.h]: No leading underscores on identifiers.

File: src/pointer-analysis/value_set_analysis.h [hunk 5]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
   }
 
-  typedef static_analysist<value_set_domaint> baset;
-
   // overloading
-  virtual void initialize(const goto_programt &goto_program);
-  virtual void initialize(const goto_functionst &goto_functions);
+  void initialize(const goto_programt &goto_program)
+  {
+    baset::initialize(goto_program);
```
Comment (reuk) [inline src/pointer-analysis/value_set_analysis.h]: Overridden methods calling base class methods is a code smell. It implies that the customization points have been chosen incorrectly. This should probably be refactored.

File: src/pointer-analysis/value_set_analysis.h [hunk 6]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  }
+  void initialize(const goto_functionst &goto_functions)
```
Comment (reuk) [inline src/pointer-analysis/value_set_analysis.h]: I think these methods are missing `override`s too.

File: src/pointer-analysis/value_set_analysis.h [hunk 7]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+      for(value_sett::valuest::const_iterator
+            v_it=value_set.values.begin();
+          v_it!=value_set.values.end();
+          v_it++)
+      {
+        xmlt &var=i.new_element("variable");
+        var.new_element("identifier").data=
+          id2string(v_it->first);
+
+#if 0
```
Comment (reuk) [inline src/pointer-analysis/value_set_analysis.h:72]: If this is never used it should probably be removed.

File: src/pointer-analysis/value_set_domain.h [hunk 8]
```diff
@@ -17,10 +17,11 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "value_set.h"
 
+template<class Value_Sett>
 class value_set_domaint:public domain_baset
 {
 public:
-  value_sett value_set;
+  Value_Sett value_set;
```
Comment (reuk) [inline src/pointer-analysis/value_set_domain.h]: More snake-camel case

File: src/pointer-analysis/value_set_domain.h [hunk 9]
```diff
@@ -44,18 +45,59 @@ class value_set_domaint:public domain_baset
...
+      break;
+
+    case END_FUNCTION:
+    {
+      value_set.do_end_function(
+        static_analysis_baset::get_return_lhs(to_l), ns);
+      break;
+    }
+
+    case RETURN:
```
Comment (reuk) [inline src/pointer-analysis/value_set_domain.h]: Add a comment to make it clear that fall-through is desired.

File: src/pointer-analysis/value_set.h [hunk 10]
```diff
@@ -242,28 +242,28 @@ class value_sett
...
-  void get_value_set_rec(
+  virtual void get_value_set_rec(
```
Comment (reuk) [inline src/pointer-analysis/value_set.h]: Again, I find it extremely unlikely that these methods are already suitable customization points. It might be better to find a more focused group of custom behaviors which can be given private virtual interfaces, and called from these (still non-virtual) methods.

**Review by reuk — CHANGES_REQUESTED** [review]

**Comment by martin-cs** [comment]
On Fri, 2017-09-22 at 13:26 +0000, Chris Smowton wrote:
> This provides sufficient hooks in value-set analysis, a CBMC core pointer-analysis, to permit downstream products such as the security analyser to create derivatives that share most of its logic.

Sounds good.

> VSA itself is currently unused, as revealed by the need to add support for DEAD statements before it can be used; this also adds unit tests to check its basic functionality has not broken in the meantime.

I don't quite recall the history but it was either written for CBMC but
we used a slightly different version or it was used in other tools;
possibly Wolverine...

> VSA is based on `static_analysist` and so will need to be ported to `ait` in the future if we wish to continue using it;

Yes; this was the last part of the "we should probably only have one
general abstract interpretation framework".

>  however this should be orthogonal to permitting it to be customised like this.

Yes.  There is also quite a bit of duplicate code in the various value
analysis variants; if you are refactoring it might be worth having a
look at that (the duplication was one of the motivations behind the
variable precision domain stuff).  Do be a little careful though because
IIRC there are some subtle hacks in this code to handle C pointer
weirdness.

Longer term there may be a question over whether configurations of the
variable precision domain might replace the functionality of
pointer_analysis/ but even if it does, it will be a while before this
code is removed.

HTH

File: src/pointer-analysis/show_value_sets.h [hunk 11]
```diff
@@ -12,10 +12,10 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_POINTER_ANALYSIS_SHOW_VALUE_SETS_H
 #define CPROVER_POINTER_ANALYSIS_SHOW_VALUE_SETS_H
 
+#include <pointer-analysis/value_set_analysis.h>
```
Comment (thk123) [inline src/pointer-analysis/show_value_sets.h:15]: Why swap this to a full include when forward declaration was sufficient?

File: src/pointer-analysis/value_set_analysis.h (hunk 3 above)
Comment (thk123) [reply src/pointer-analysis/value_set_analysis.h]: I don't think adopting a different naming convention is a good way of disambiguating it from the existing `value_sett`. Perhaps `selected_value_sett`? Perhaps there is a better term. `value_set_domaint`?

File: src/pointer-analysis/value_set_analysis.h [hunk 12]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
 
   void convert(
     const goto_programt &goto_program,
     const irep_idt &identifier,
-    xmlt &dest) const;
+    xmlt &dest) const
+  {
+    source_locationt previous_location;
+
+    forall_goto_program_instructions(i_it, goto_program)
```
Comment (thk123) [inline src/pointer-analysis/value_set_analysis.h:47]: Nit: while in the area a range based for might be better here?

File: src/pointer-analysis/value_set_analysis.h [hunk 13]
```diff
@@ -14,32 +14,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+      if(location.is_nil() || location.get_file().empty())
+        continue;
+
+      // find value set
+      const value_sett &value_set=(*this)[i_it].value_set;
+
+      xmlt &i=dest.new_element("instruction");
+      i.new_element()=::xml(location);
+
+      for(value_sett::valuest::const_iterator
```
Comment (thk123) [inline src/pointer-analysis/value_set_analysis.h:63]: Nit: the formatting on this for loop is a bit hard to follow. You might consider using `auto` to get each statement on one line or at least aligning each statement

File: src/pointer-analysis/value_set_analysis.h [hunk 14]
```diff
@@ -48,10 +106,12 @@ class value_set_analysist:
     const exprt &expr,
     value_setst::valuest &dest)
   {
-    (*this)[l].value_set.get_value_set(expr, dest, ns);
+    (*this)[l].value_set.get_value_set(expr, dest, baset::ns);
   }
 };
 
+typedef value_set_analysis_baset<value_sett> value_set_analysist;
```
Comment (thk123) [inline src/pointer-analysis/value_set_analysis.h]: This seems like a weird place to introduce a typedef? If really needed in the header file outside of the class, perhaps put it above the class?

File: unit/CustomVSATest.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
+
+public class CustomVSATest {
```
Comment (thk123) [inline unit/CustomVSATest.java]: Could this go in the same folder as the test to avoid cluttering `unit`

File: unit/pointer-analysis/custom_value_set_analysis.cpp [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+    bool model_init_failed=
+      initialize_goto_model(goto_model, command_line, null_output);
+
+    namespacet ns(goto_model.symbol_table);
+
+    // Fully inline the test program, to avoid VSA conflating
+    // constructor callsites confusing the results we're trying to check:
+    goto_function_inline(goto_model, TEST_FUNCTION_NAME, null_output);
+
+    REQUIRE(!model_init_failed);
```
Comment (thk123) [inline unit/pointer-analysis/custom_value_set_analysis.cpp]: Any reason not to do this REQUIRE immediately after `initalize_goto_model`?

**Review by thk123 — COMMENTED** [review]
This looks good - just a few minor corrections. I also agree with @reuk that the interface could probably be cleaned up so the public methods aren't virtual. I will read up on Value Set Analysis so that I can validate that the tests covers suitable cases - but they look very clean!

File: unit/pointer-analysis/custom_value_set_analysis.cpp [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+/// An example customised value_sett. It makes a series of small changes
+/// to the underlying value_sett logic, which can then be verified by the
+/// test below:
+/// * Writes to variables with 'ignored' in their name are ignored.
+/// * All writes are made weak by forcing `add_to_sets` to true.
+/// * Never propagate values via the field "never_read"
+/// * Adds an ID_unknown to the value of variable "maybe_unknown", and
+///     to the possible referees of fields named unknown_field_ref and
+///     variables named unknown_global_ref
+class custom_value_sett:public value_sett
```
Comment (thk123) [inline unit/pointer-analysis/custom_value_set_analysis.cpp]: You might consider pulling this class out into a separate file with just the above documentation left in the test file since it isn't directly relevant to the test. Definitely not a requirement, so do disregard if you disagree.

**Review by thk123 — COMMENTED** [review]
The tests seem good. I'm tempted to say the methods that you don't provide a custom implementation for probably shouldn't be virtual

File: src/pointer-analysis/value_set.h (hunk 2 above)
Comment (smowton) [reply src/pointer-analysis/value_set.h:259]: The virtual override points definitely need defer-to-base I'm afraid

File: src/pointer-analysis/value_set_analysis.h (hunk 7 above)
Comment (smowton) [reply src/pointer-analysis/value_set_analysis.h:72]: Shrug, that goes for the entire XML interface. I don't know who (if anyone) uses it and if they use this sometimes, so I'm leaving it alone.

File: src/pointer-analysis/show_value_sets.h (hunk 11 above)
Comment (smowton) [reply src/pointer-analysis/show_value_sets.h:15]: `value_set_analysist` is a template specialisation now; `class value_set_analysist` won't work as the forward-decl will clash with the actual definition. There might be a way around this but I don't know it.

**Comment by martin-cs** [comment]
> Shrug, that goes for the entire XML interface. I don't know who (if anyone) uses it and if they use this sometimes, so I'm leaving it alone. 

The XML interface was developed for a customer (I forget exactly who)
and I think there are several people who use it. Any decision to
depreciate it should be done across the whole code-base.

File: src/pointer-analysis/value_set_analysis.h (hunk 12 above)
Comment (smowton) [reply src/pointer-analysis/value_set_analysis.h:47]: True, re: this and below though I'll stick to just moving the code without many changes and perhaps clean up later.

**Comment by smowton** [comment]
@reuk @thk123 all changes applied except as noted.

File: src/pointer-analysis/value_set_analysis.h [hunk 18]
```diff
@@ -14,32 +14,80 @@ Author: Daniel Kroening, kroening@kroening.com
...
+#include <util/xml.h>
 
 #include "value_set_domain.h"
 #include "value_sets.h"
+#include "value_set.h"
 
 class xmlt;
 
-class value_set_analysist:
+template<class DomainValueSet>
```
Comment (thk123) [inline src/pointer-analysis/value_set_analysis.h]: Still think looks a bit awkward to have this non-conforming to name convention (pointing out in case this isn't intentional, not to start a flame war :stuck_out_tongue:)

**Review by thk123 — APPROVED** [review]
The code and tests look great, I would feel more conformable if someone who was already familiar with the behaviour of these analyses checked the tests were sufficient.

**Comment by smowton** [comment]
Changed names again to VST, believe everything is applied now

File: src/pointer-analysis/value_set.h [hunk 19]
```diff
@@ -287,6 +288,54 @@ class value_sett
...
+    bool add_to_sets);
+
+  /// Subclass customisation point for applying code to this domain:
+  virtual void apply_code_rec(
+    const codet &code,
+    const namespacet &ns);
+
+  /// Expression simplification function; by default, plain old
+  /// util/simplify_expr, but can be customised by subclass.
+  expr_simplifiert simplifier;
```
Comment (reuk) [inline src/pointer-analysis/value_set.h]: Make private, and provide a protected method which allows this to be called (but nothing else)

File: src/pointer-analysis/value_set.h [hunk 20]
```diff
@@ -24,8 +24,20 @@ class namespacet;
...
-  value_sett():location_number(0)
+  value_sett():
+    location_number(0),
+    simplifier(default_simplifier)
+  {
+  }
+
+  explicit value_sett(expr_simplifiert simplifier):
+    location_number(0),
+    simplifier(simplifier)
```
Comment (reuk) [inline src/pointer-analysis/value_set.h]: Should `std::move` the simplifier

File: unit/pointer-analysis/custom_value_set_analysis.cpp [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+
+#include <util/config.h>
+#include <langapi/mode.h>
+#include <goto-programs/initialize_goto_model.h>
+#include <goto-programs/goto_inline.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/java_types.h>
+#include <pointer-analysis/value_set_analysis.h>
+
+/// Counts calls to our custom simplifier, registered above:
```
Comment (reuk) [inline unit/pointer-analysis/custom_value_set_analysis.cpp]: Below...

File: unit/pointer-analysis/custom_value_set_analysis.cpp [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+      }
+    }
+  }
+};
+
+typedef value_set_analysis_baset<test_value_sett> test_value_set_analysist;
+
+#define TEST_PREFIX "java::CustomVSATest."
+#define TEST_FUNCTION_NAME TEST_PREFIX "test:()V"
+#define TEST_LOCAL_PREFIX TEST_FUNCTION_NAME "::"
```
Comment (reuk) [inline unit/pointer-analysis/custom_value_set_analysis.cpp:146]: You could use `static constexpr const char *` here and avoid the evil preprocessor

File: src/pointer-analysis/value_set_domain.h [hunk 23]
```diff
@@ -44,17 +45,59 @@ class value_set_domaint:public domain_baset
...
-    locationt from_l,
-    locationt to_l);
-
   virtual void get_reference_set(
     const namespacet &ns,
     const exprt &expr,
     value_setst::valuest &dest)
   {
-    value_set.get_reference_set(expr, dest, ns);
+    value_set.read_reference_set(expr, dest, ns);
```
Comment (reuk) [inline src/pointer-analysis/value_set_domain.h]: In general, if you are specifying interface requirements on a templated item, use non-member rather than member functions (i.e. `read_reference_set(value_set, expr, dest, ns)`). This allows pre-existing types to be used, without altering their definitions, as the template requirements can be fulfilled by providing external, non-member functions. If necessary, these can even be local to a translation unit, allowing template clients in different locations to use different implementations of these wrapper/interface functions.

File: unit/pointer-analysis/custom_value_set_analysis.cpp (hunk 22 above)
Comment (smowton) [reply unit/pointer-analysis/custom_value_set_analysis.cpp:146]: AFAIK I can't concatenate them as is done here, so I'd prefer to leave as-is

File: src/pointer-analysis/value_set_domain.h (hunk 23 above)
Comment (smowton) [reply src/pointer-analysis/value_set_domain.h]: Don't follow this, could I have an example?

**Comment by smowton** [comment]
Comments 1-3 done, 4 declined, 5 I don't understand, please give an example?

File: src/pointer-analysis/value_set_domain.h (hunk 23 above)
Comment (reuk) [reply src/pointer-analysis/value_set_domain.h]: Let's say I have a type that already exists - maybe it's a `std::vector` or something where we're not able to change the implementation. Now, if you were to use this type as your template parameter, you would probably find that it doesn't have the right member functions to satisfy the interface implicitly specified by the template class. The problem is that there's nothing you can do now - you can't modify the class definition to add the right methods.

If you use non-member functions inside your template, you have more flexibility to define appropriate implementations of these functions without modifying the original class. This also gives more flexibility when refactoring. At the moment, if you changed a method in one of your classes that's used as a `VST`, you would have to go and modify this template, too. However, if you use non-member functions you can just modify the non-member passthrough function to call the modified member function (one change), without having to make changes throughout the template class (potentially several changes).

This is about separating the general interface for a class (its member functions) from the specific interface required for a particular template class (non-member functions).

Comment (smowton) [reply src/pointer-analysis/value_set_domain.h]: OK, done this this time. In general I think this sort of layer of indirection should be introduced on demand though, as the case where we need it is fairly unlikely (maybe 10% chance of needing this over the life of the code, considering the template parameters are internal code, and the cost of introducing the indirection scales with the size of the implied interface, not the number of subclasses already providing it.

**Comment by smowton** [comment]
Change 5 now applied too.

**Review by reuk — APPROVED** [review]

**Comment by polgreen** [comment]
Hi, 
It would be helpful if in future, when large-ish changes like this are introduced, the old api could be kept in develop in addition to the new one at least for a short time, please. Then people who depend on the existing api have a bit of time to alter their code to fit the new api, whilst still being able to rebase against develop to pull in other new changes.
Thanks!

**Comment by smowton** [comment]
1. Sorry, didn't think anybody was using this code except for goto-instrument.
2. What API break? `value_set_analysist` and `value_sett` *should* be exactly as-was
3. Daniel has requested this reverted, which will happen once CI has passed

**Comment by martin-cs** [comment]
On Mon, 2017-10-16 at 01:28 -0700, Chris Smowton wrote:
> 1. Sorry, didn't think anybody was using this code except for goto-instrument.
It should be used in symex and in the dependency graph / slicer at a
minimum.  As a generally point, we can't entirely assume that something
is unused just because it is not used in the main CBMC repository.

Also, where possible, introducing the new API, depreciating the old one
and then removing it later is A Good Thing.  See the ai.h /
static_analysis.h divide.

**Comment by polgreen** [comment]
I'm using value_set_domaint, which has changed quite a bit. I've fixed my stuff now, just it would be nice if there's a bit of a period of overlap for future changes like this.

**Comment by smowton** [comment]
Ah ok, so would my writing typedef value_set_domaint templated_value_set_domaint<value_sett> have prevented problems, or were there further problems with value_set_domaint's internals?

**Comment by polgreen** [comment]
You would also need to keep the names of the functions within templated_value_set_domaint<value_sett> the same as the functions within value_set_domaint (e.g., get_value_set is now read_value_set, and a new get_value_set has been introduced)

**Comment by smowton** [comment]
Ah crap, sure, that was because the 4-operand and 3-operand versions of `get_value_set` were needlessly overloads, which led to awkwardness when subclassing `value_sett` as the 3-op version was hidden by overriding the 4-op one. I can make an alias instead. FWIW there isn't a new get_value_set, only the old 4-op version probably showing through where existing code expects the 3-op one. Again the cause was thinking value_set_domaint was a "private" implementation artefact of value_set_analysist.

In any case this is all now reverted per @kroening.

**Comment by smowton** [comment]
@polgreen made a replacement to this PR here: https://github.com/diffblue/cbmc/pull/1499

Since this broke your stuff the last time around, would you mind checking whether you can merge that PR into your cbmc branch without causing undue mayhem?

**Comment by polgreen** [comment]
That's great, thanks!
